File: 416_Partition_Equal_Subset_Sum.cpp
#include<bits/stdc++.h>

using namespace std;

class Solution {
public:
    bool canPartition(vector<int>& nums) {
        int n = nums.size();
        int sum = accumulate(nums.begin(), nums.end(), 0);
        if (n <=1 || sum % 2 != 0) return false;
        int t = sum / 2;
        
        vector<bool> dp(t + 1, false); dp[0] = true;
        
        for (int num : nums){
            for (int j = t; j >= num; --j){
                dp[j] = dp[j - num] || dp[j];
            }
        }
        
        return dp[t];
    }
};

int main(){
    vector<int> nums{5, 1, 5, 11};
    Solution* solution = new Solution();
    cout << solution->canPartition(nums) << endl;
    return 0;
}

File: 416_Partition_Equal_Subset_Sum.py
class Solution(object):
    def canPartition(self, nums):
        """
        :type nums: List[int]
        :rtype: bool
        """
        def backtrack(nums, t, seen):
            if t in seen: return seen[t]
            if t == 0:
                seen[t] = True
                return True

            seen[t] = False
            for i in range(len(nums)):
                if t - nums[i] >= 0:
                    seen[t] = seen[t] | backtrack(nums[i+1:], t - nums[i], seen)
                    if seen[t]:
                        return True
            return seen[t]

        s = sum(nums)
        if s % 2: return False
        return backtrack(nums, s // 2, {})

s = Solution()
print(s.canPartition([1,5,11,5]))
print(s.canPartition([1,2,3,5]))